### ISSUE TITLE
Add support for authentication 

### ISSUE THREAD

**Issue by patrick91** [strawberry-graphql/strawberry#830]
I've opened a discussion on this, but I think it might be worth making an issue related to authentication. 

I'd say we can split this work into multiple chunks, but let's think about what we want in ideal world.

- [ ] We should be able to use different ways of authentication[1]
- [ ] We should be able to have pluggable user types
- [ ] We should be able to generate login, logout and signup mutations
- [ ] We should be able to have pluggable input types and return types for these mutations[2]
- [ ] We should be able to support django but also any other framework with not much code

[1] JWT with cookies, JWT in header, plain cookies for sessions (basic django auth)
[2] we shouldn't force users to use our preferred way of error handling

All of the above is up for discussion by the way :) haven't really found any other library that do authentication built-in. For Graphene we have these two:

- https://github.com/PedroBern/django-graphql-auth
- https://github.com/flavors/django-graphql-jwt

**Comment by kristofgilicze** [comment]
I can see why this make sense for the django integration.
Common features working out-of-the-box (with a few lines of settings.py) is expected there.
[DRF style API](https://www.django-rest-framework.org/api-guide/authentication/
) would be pretty familiar to most django devs.

When using a minimal framework like Flask, this might feel a bit unusual since Flask has no built in auth.
Sanic is the same,  just a few example in the [docs](https://sanicframework.org/en/guide/how-to/authentication.html#server.py). 
So I believe this should be only for the django and the 'django-like' integrations.

Shipping a first-party `strawberry-graphql/auth` ext is also something to consider, **though I would the prefer the built-in approach.**

**Comment by jkimbo** [comment]
I agree with @kristofgilicze I think an authentication integration makes sense with Django because it's provided by the framework. Maybe we can build something into the Django integration @la4de ? 

Something like that Sanic guide would be good for the other frameworks.

**Comment by patrick91** [comment]
I wonder though, if we build support for django, wouldn't be "easy" to make it pluggable and support multiple frameworks. The mutations and queries would be the same everywhere, no? 

The implementation would change, so maybe we can provide some hooks for that?

**Comment by la4de** [comment]
We have preliminary support already in `strawberry-graphql-django` package. It's still in very early stage but you can already generate graphql type from user model and login / logout with given auth mutations.

**Comment by patrick91** [comment]
@la4de that's awesome, so maybe we can keep this in strawberry-django and then when it is mature we can think of making more generic, what do you all think?

**Comment by kristofgilicze** [comment]
core implementation should be reused between integrations.

The django way to configure stuff is in the `settings.py`.
I think it is already a little weird that I set the schema on the view.

**Comment by patrick91** [comment]
> I think it is already a little weird that I set the schema on the view.

It's definitely different than django, but it allows to have multiple GraphQL views in one project (we do that at work for example).

**Comment by kristofgilicze** [comment]
If you have multiple views, then I agree the 'global' setting does not make sense.

I had a closer look at how DRF handles the auth classes.

... there is also an option to set it on a per-view basis.

What I described before:
 in the settings.py you actually set / override the **default** auth classes.

see:
https://www.django-rest-framework.org/api-guide/authentication/#setting-the-authentication-scheme

**Comment by joeydebreuk** [comment]
> @la4de that's awesome, so maybe we can keep this in strawberry-django and then when it is mature we can think of making more generic, what do you all think?

I think that it should remain framework specific. There are too many things to cover, too many different frameworks and too many different approaches. While within the context of a framework for a specific auth method, its really simple, you can see [here](https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/strawberry_django/mutations/auth.py).

I'd suggests to make it really simple for integrations (eg strawberry-graphql-django) and for users to:

1. Add something like "auth" to "context"
2. Use this context for permission classes.

**Comment by joeydebreuk** [comment]
@patrick91 I feel like we had a similar discussion regarding adding framework specific code to strawberry. I suppose its a matter of philosophy and defining the scope of strawberry.

I'm sure you can add some abstraction on the strawberry level, but I don't see how it would add a lot of value. Just more complexity. And more docs to write.

> We should be able to use different ways of authentication[1]

Already possible, and quite simple to do. Some documentation would help though.

> We should be able to have pluggable user types

Why should strawberry be are of user types?

> We should be able to generate login, logout and signup mutations

[These](https://github.com/strawberry-graphql/strawberry-graphql-django/blob/main/strawberry_django/mutations/auth.py) mutations are very simple, I don't see the point in doing this.

> We should be able to have pluggable input types and return types for these mutations[2]

Defining these inputs and returns will be most of the code one would need for a mutation in the first place

> We should be able to support django but also any other framework with not much code

Maybe its an idea to just provide a guide for each framework?

> even django has pluggable auth backends :)

Django has everything though :smile: 

In short, I would say that adding a few guides for setting up strawberry for frameworks is better than adding framework-specific code to strawberry.

**Comment by patrick91** [comment]
> In short, I would say that adding a few guides for setting up strawberry for frameworks is better than adding framework-specific code to strawberry.

Could be! It's just that I see auth something used a lot and quite critical (auth should be implemented well and should secure), so it would be nice if we can help with that as a framework.

I'll see if I can come up with some examples of what I am thinking 😊

**Comment by taybenlor** [comment]
Weighing in here, I'm personally not at all interested in Django integrations so this discussion seems a bit too "heavyweight" for me. I'm providing a GraphQL alternative to a FastAPI endpoint, which is conceptually similar to people discussing Flask.

I'd like to be able to share some code between the two, but currently I'm most interested in Strawberry providing tools like hooks, helpers, and mixins that would make Authentication easier. Ideally I would be free to define my own mutations and permissions on queries, but with some deeper strawberry integration.

For example a solution to me might be for Strawberry to simply provide tools that work like the current Permission classes. Then make it my responsibility to provide login and signup mutations.

```
@strawberry.authenticated(any=[HeaderAuthentication, SimpleHTTPAuthentication, CookieAuthentication])
@strawberry.type
class SomeQuery:
   @strawberry.field
   def some_resolver(self, info) -> str:
       user = info.user

@strawberry.authenticated(all=[CookieAuthentication, AdminAuthentication])
class SomeAdminQuery:
    @strawberry.field
    def some_resolver(self, info) -> str:
        admin = info.user
```

In this scenario the `any` argument and the `all` arguments allow for authentication that can be structured as alternatives (any) or layered checks (all). The info object is expanded to have a configurable/settable property called `user`.

I'm not suggesting this be the exact implementation nor am I married to any particular approaches here. Just making suggestions that may be simpler right now as a way to progress this ticket without thinking about Django.

**Comment by karolzlot** [comment]
It would be great if strawberry would be able to reuse FastAPI's `Depends(deps.get_current_active_user)` and `Depends(deps.get_current_active_superuser)` as demonstrated in these links:

https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/api/api_v1/endpoints/items.py

https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/api/api_v1/endpoints/users.py

This way is super comfortable to use!

(after you go to the links above, you can click on those method names to see how they are defined)

---

Maybe this information will be helpful in implementing it (it is about graphene): 

https://stackoverflow.com/questions/61930695/how-to-get-user-auth-info-in-fastapi-when-using-app-add-route-for-graphql

**Comment by jkimbo** [comment]
@karolzlot I personally like the dependency injection API that FastAPI has and I'd like to see something similar implemented in Strawberry. It's unlikely that we would be able to use FastAPI dependencies directly though because it's quite paradigm. For example in the `get_current_active_user` dependency that you linked it raises an HTTP error if there isn't a current user which is not something we would want to do in GraphQL.

**Comment by nrbnlulu** [comment]
So i am trying to migrate [django-graphql-auth](https://github.com/PedroBern/django-graphql-auth ) to strawberry and they have that thing in mutations that they initialize the input dynamically by supplying a dict or list with the desired  fields and i have managed to do the same with strawberry like in the attached file.

```python
def make_dataclass_helper(required: dict | list, non_required: dict | list):
    res_req = []
    res_non_req = []

    if isinstance(required, dict):
        for key in required:
            res_req.append((key, required[key]))

    elif isinstance(required, list):
        for key in required:
            res_req.append((key, str))

    if isinstance(non_required, dict):
        for key in non_required:
            res_non_req.append((key, typing.Optional[non_required[key]], dataclasses.field(default=None)))

    elif isinstance(non_required, list):
        for key in non_required:
            res_non_req.append((key, typing.Optional[str], dataclasses.field(default=None)))


    return res_req + res_non_req


class DynamicInputMixin:
    """
    A class that knows how to initialize strawberry input

    as dict { input_name: input_type }
    as list [input_name,] -> defaults to String

    #usage:
        class SomeMutation(DynamicInputMixin, MutationMixin):
                class _meta:
                    _inputs = [some, non, required, fields]
                    _required_inputs = {some:str, required:int, fields:list}

                def resolve_mutation(self, input_):
                    logic...

    """

    def __init_subclass__(cls, **kwargs):
        _inputs = cls._meta._inputs
        _required_inputs = cls._meta._required_inputs
        if not isinstance(_inputs, dict | list):
            raise WrongUsage(f'dynamic inputs can be list or dict not{type(_inputs)}')
        if not isinstance(_required_inputs, dict | list):
            raise WrongUsage(f'dynamic required inputs can be list or dict not{type(_required_inputs)}')

        dc = dataclasses.make_dataclass('Input',
                                        make_dataclass_helper(_required_inputs, _inputs),
                                        namespace={
                                            'get': lambda self, attr: self.__getattribute__(attr)
                                        }
                                        )

        inputs = strawberry.input(dc)
        setattr(cls._meta, 'inputs', inputs)
        super().__init_subclass__()


class DynamicPayloadMixin:
    """
        A class that knows how to initialize strawberry
        payload from a list or dict

        as dict { payload_name: payload_type }
        as list [input_name,] -> defaults to String

        #usage:
            class SomeMutation(DynamicInputMixin, OutputMixin, MutationMixin):
                    _outputs = [some, non, required, fields]
                    _required_outputs = {some:str, required:int, fields:list}

                    def resolve_mutation(self, input_):
                        logic...
    """

    def __init_subclass__(cls, **kwargs):
        _outputs = getattr(cls._meta, '_outputs', None)
        _required_outputs = getattr(cls._meta, '_required_outputs', None)
        if _outputs or _required_outputs:
            if not isinstance(_outputs, dict | list | None):
                raise WrongUsage(f'dynamic outputs can be list or dict not{type(_outputs)}')
            if not isinstance(_required_outputs, dict | list | None):
                raise WrongUsage(f'dynamic required outputs can be list or dict not{type(_required_outputs)}')

            dc = dataclasses.make_dataclass('outputs',
                                            make_dataclass_helper(_required_outputs, _outputs),
                                            namespace={
                                                'get': lambda self, attr: self.__getattribute__(attr)
                                            },
                                            order=False
                                            )

            class Outputs(dc, OutputInterface):
                ...

            outputs = strawberry.type(Outputs)
        else:
            outputs = strawberry.type(OutputInterface)

        setattr(cls, 'output', outputs)
        super().__init_subclass__()


class MutationMixin:
    def __init_subclass__(cls, **kwargs):
        @strawberry.mutation
        def mutation(input_: cls._meta.inputs) -> cls.output:
            return cls.resolve_mutation(input_)

        setattr(cls, cls.__name__.lower().strip('mutation'), mutation)


@strawberry.type
class RegisterMutation(DynamicInputMixin, DynamicPayloadMixin, MutationMixin):
    class _meta:
        _required_inputs = ['blip'] 
        _inputs = ['testOUT','testIN2']

        _required_outputs = ['success']
        _outputs = ['testOut', 'message']



    @classmethod
    def resolve_mutation(cls, input_):
        """
        logic goes here
        """
        blip = input_.get('blip')
        if blip:
            return cls.output(success=True, message='blop')
```

**Comment by ifaizanali** [comment]
Is there a way to use logical operators in strawberry permissions classes like:
`permission_classes = [IsAuthenticates, ((IsOwner and IsMember) or IsAdmin)]`
something like that ^^^^^

**Comment by yhdelgado** [comment]
Is there any advance on this topic? I'm seeking a practical solution to validate a third-party token on FastAPI+Strawberry.

**Comment by elefher** [comment]
is there any update on this topic? Some of the implementations mentioned above would be very useful

**Comment by patrick91** [comment]
@elefher what implementations would like to see the most?

**Comment by danielmcquillen** [comment]
+1 for a simple authorization token in header approach similar to what DRF provides ( and slightly lighter weight than JWT ).
